### DISCUSSION TITLE
Next.js – on initial page load, the slidesPerView settings appear to be ignored, resulting in content flicker

### DISCUSSION THREAD

**Discussion by sebpowell** [nolimits4web/swiper#6785] [Q&A]
Hi there, thank you for this great library. I've tried a few, and this is hands down the best one.

I was hoping you might be able to advise me on the following issue that I've encountered when using it with Next.js (13.4.2).

On initial page load, it would appear all `slidesPerView` settings are being ignored, please see the below video. I would like to avoid the 'flicker' that you can see in the video – is this possible, and if so, what would be the right way to go about it? I have tried a few different things, including checking that the component has mounted before showing the slider, but no luck as of yet. 

https://github.com/nolimits4web/swiper/assets/1786366/2de23af1-40b1-44f7-8f7d-2182a1d4fff1

The Codesandbox for the above video is available here: https://codesandbox.io/p/sandbox/cool-goldberg-3t629p

Any advice would be greatly appreciated.

**Comment by AbdushukurRasulov** [comment]
@sebpowell 
did you fix

**Comment by sebpowell** [comment]
@AbdushukurRasulov – no sadly not. Are you having the same problem?

**Comment by AbdushukurRasulov** [comment]
Yes, the same problem

**Comment by sebpowell** [comment]
@AbdushukurRasulov – it's a problem with other libs too, here's a proposed solution for KEEN: https://github.com/rcbyr/keen-slider/issues/16. I also found this that I might try building: https://medium.com/@yi.zhuang1996/building-a-react-carousel-component-that-rocks-with-server-side-rendering-support-part-1-aba5ccfae76e

**Comment by sebpowell** [comment]
@AbdushukurRasulov btw, the above seems to be based on this: https://github.com/YIZHUANG/react-multi-carousel/tree/master – which apparently supports SSR, although I tried the demo and was still getting a flicker.

**Reply by AbdushukurRasulov** [reply]
@sebpowell tried with another library, the same issue.
please, ping here if you found solution.
Thank you ))

**Reply by sebpowell** [reply]
@AbdushukurRasulov – did you try the one above? Apparently, that one should work, because it works with SSR (at least, that's what it says) – I haven't had a chance to try it though.

**Comment by msobkyy** [comment]
same behavior

**Comment by bank188950** [comment]
I put this code in component. Very confused. What does it have to do with it?
--- Fixbug ---
const initialTheme: string = localStorage.getItem('theme') || 'light';

**Comment by dwalker93** [comment]
First you have to check whether the dom fully loaded.
use a useEffect and useState in Swiper componet like this.
It's a good user experience return a skeleton rather than returning null while dom loading.

```
const SwiperSlide = () => {
const [load, setLoad] = useState(false);
  useEffect(() => {
    setLoad(true);
  }, []);

  if (!load) return null;
  return (
       <Swiper>
            <SwiperSlide>
             ..............
            </SwiperSlide>
       </Swiper>
   )
}
```

**Comment by ehmasuk** [comment]
i found this as a solution and it works for me. I don't know how does it work, but it fixed the issue.
---Solution---

const initialTheme = localStorage.getItem('theme') || 'light';

add this code on top of the component where you are using the slider, thats it. It should solve the issue

**Reply by nctay** [reply]
Yep, it works. Definitely some hack with client rendering, although i use 'use client' directive everywhere.

**Reply by memoday** [reply]
This is awesome

**Reply by Sergiobmrq** [reply]
Working on dev environment, build throws an error

**Comment by mahmoudAcm** [comment]
@sebpowell  here is my [workaround ](https://github.com/nolimits4web/swiper/discussions/7578)

**Comment by abolfazl-zare** [comment]
i have same problem

**Comment by codebasecodebase** [comment]
None of the methods provided above help.

**Reply by mahmoudAcm** [reply]
here is my [workaround ](https://github.com/nolimits4web/swiper/discussions/7578)

**Comment by MohammadSoleimanikia** [comment]
### Swiper.js does not initialize correctly on the first load

I’ve tried all of the solutions mentioned above, but the issue still occurs.

From what I’ve found, the problem seems to happen because the Swiper JavaScript code has not been loaded/initialized yet when the component is rendered or when the component using Swiper is not fully mounted.

One possible solution is to create a small `useMount` hook that detects whether the component has been mounted on the client.

While the component is not mounted, we can render a skeleton/loading placeholder instead of initializing Swiper. Once `mounted` becomes `true`, we can render the Swiper component.

### Example approach

```ts
function useMount() {
  const [mounted, setMounted] = useState(false);

  useEffect(() => {
    setMounted(true);
  }, []);

  return mounted;
}
```

Then in the component:

```tsx
function MySwiper() {
  const mounted = useMount();

  if (!mounted) {
    return <SwiperSkeleton />;
  }

  return (
    <Swiper>
      {/* slides */}
    </Swiper>
  );
}
```

This approach makes sure that Swiper is only rendered after the component has been mounted on the client, which seems to prevent the first-load initialization issue.

I’m sharing this as a possible workaround for anyone experiencing the same first-load Swiper issue. I’d also be interested to know if there is a more official or recommended way to handle this in Swiper.js.
